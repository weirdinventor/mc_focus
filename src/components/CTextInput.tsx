// /components/CTextInput.tsx (CORRIGÉ AVEC forwardRef)

import React, { ComponentType, useMemo, useRef, useState } from 'react';
import { FieldError } from 'react-hook-form';
import { CText } from './CText'; // Assurez-vous que ce composant existe et est adapté au web
import { t } from "i18next";

// --- Mocks et Styles (inchangés) ---
const Colors = { /* ... vos couleurs ... */ };
const TestIDs = { tests: { CTextInputBottomText: 'ctext-input-bottom-text' } };
const componentStyles = `/* ... vos styles ... */`;

// --- Type Definitions for Web (MODIFIÉES) ---
type TextInputStatuses = 'disabled' | 'error' | 'focused';
type AccessoryProps = { state?: TextInputStatuses };

// On enlève `inputRef` d'ici, car il sera géré par `forwardRef`
export type CustomTextInputProps = {
  state?: TextInputStatuses;
  placeholderText?: string;
  error?: FieldError;
  LeftAccessory?: ComponentType<AccessoryProps>;
  RightAccessory?: ComponentType<AccessoryProps>;
  bottomText?: string;
  bottomTextColor?: string;
  wrapperClassName?: string;
} & Omit<React.InputHTMLAttributes<HTMLInputElement>, 'placeholder'>;


/**
 * Un composant de saisie de texte personnalisable qui propage la ref.
 * Enveloppé dans React.forwardRef.
 */
export const CTextInput = React.forwardRef<HTMLInputElement, CustomTextInputProps>(
  ({
    state,
    placeholderText,
    LeftAccessory,
    RightAccessory,
    bottomText,
    bottomTextColor = 'red',
    error,
    className,
    wrapperClassName,
    style,
    disabled: propsDisabled,
    ...textInputProps
  }, ref) => { // La ref est maintenant le deuxième argument !

    const [isFocused, setIsFocused] = useState(false);

    const disabled = propsDisabled || state === 'disabled';
    const hasError = !!error || state === 'error';
    const placeholderString = (placeholderText && t(placeholderText));

    const focusTextInput = () => {
      if (disabled) return;
      // Pour faire le focus, on peut vérifier si la ref est un objet
      if (ref && typeof ref !== 'function') {
        ref.current?.focus();
      }
    };

    // --- Logique des classes CSS (inchangée) ---
    const linearOuterClasses = useMemo(() => [/* ... */].filter(Boolean).join(' '), [isFocused, hasError]);
    const wrapperClasses = useMemo(() => [/* ... */].filter(Boolean).join(' '), [isFocused, wrapperClassName]);
    const inputClasses = useMemo(() => [/* ... */].filter(Boolean).join(' '), [hasError, className]);
    const currentAccessoryState = isFocused ? 'focused' : state;

    return (
      <>
        <style>{componentStyles}</style>
        <div className="ctext-input-pressable-wrapper" onClick={focusTextInput}>
          <div className={linearOuterClasses} style={style}>
            <div className={wrapperClasses}>
              {!!LeftAccessory && <LeftAccessory state={currentAccessoryState} />}
              <input
                // ICI : La ref est directement attachée à l'élément input.
                ref={ref}
                placeholder={placeholderString}
                {...textInputProps}
                onFocus={() => setIsFocused(true)}
                onBlur={() => setIsFocused(false)}
                disabled={disabled}
                className={inputClasses}
              />
              {!!RightAccessory && <RightAccessory state={currentAccessoryState} />}
            </div>
          </div>
          
          {error?.message && hasError && !isFocused && (
            <CText
              data-testid={TestIDs.tests.CTextInputBottomText}
              size="sm_extraBold"
              style={{ color: bottomTextColor }}
              className="ctext-input-bottom-text"
            >
              {error?.message ?? bottomText}
            </CText>
          )}
        </div>
      </>
    );
  }
);