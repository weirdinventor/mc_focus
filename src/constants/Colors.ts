const pallete = {
  white: '#FFFFFF',
  black: '#000000',
  transparent: 'transparent',

  white80: '#FFFFFF80',
  whiteSand50: '#F5F5F4',
  wildSand200: '#DDDDDA',

  black30: '#00000030',
  black40: '#00000040',
  black60: '#00000060',
  black90: '#00000090',

  primaryWhite: '#FFFFFF',
  primaryBlack: '#01080a',

  deepRed: '#FF0000',
  red: '#CF6464',
  deepGreen: '#24FF00',
  green: '#4BB543',

  black1: '#202020',
  grey1: '#e3e3e3',
  grey2: '#424242',
  grey3: '#E3DEEA',
  grey4: '#F0F0EE',
  grey5: '#56564E',
  grey6: '#9D9C93',
  grey7: '#E9E9E8',
  grey8: '#818176',
  grey9: '#696860',
  grey10: '#D9D9D9',
  grey11: '#C0BFB9',

  seanceWhite: '#FDF2E9',
  seance100: '#FEF7ED',
  seance400: '#F4B951',
  seance400_50: '#F4B95150',
  seance600: '#E79C1C',
  seance700: '#E79C1C',
  seance800: '#D18B15',
  seance900: '#405c57ff',
  seance950: '#E79C1C',
  sharpPink: '#E79C1C',
  wildSand950: '#2A2A27',

  iconInactive: '#9C9C9C',
  iconActive: '#405c57ff',
} as const;

export const Colors = {
  ...pallete,
  backgroundWhite: pallete.whiteSand50,
  backgroundBlack: pallete.primaryBlack,

  seperator: pallete.primaryBlack,

  textInput: {
    light: {
      main: pallete.grey7,
    },
    dark: {
      main: pallete.primaryWhite,
    },
  },

  texts: {
    light: {
      black: pallete.primaryBlack,
      white: pallete.primaryWhite,
      white80: pallete.white80,
      grey: pallete.grey8,
      grey5: pallete.grey5,
      grey6: pallete.grey6,
      grey9: pallete.grey9,
      pink: pallete.sharpPink,
      pink700: pallete.seance700,
      deepRed: pallete.deepRed,
      seance800: pallete.seance800,
      grey11: pallete.grey11,
      wildSand950: pallete.wildSand950,
    },

    dark: {
      black: pallete.primaryWhite,
      white: pallete.primaryBlack,
      white80: pallete.white80,
      grey: pallete.grey8,
      grey5: pallete.grey5,
      grey6: pallete.grey6,
      grey9: pallete.grey9,
      pink: pallete.sharpPink,
      pink700: pallete.seance700,
      deepRed: pallete.deepRed,
      seance800: pallete.seance800,
      grey11: pallete.grey11,
      wildSand950: pallete.wildSand950,
    },
  },

  buttons: {
    primaryBorder: pallete.grey3,
    whiteBorder: pallete.white,
    light: {
      primary: pallete.white,
      secondary: pallete.grey4,
      colored: pallete.transparent,
      transparent: pallete.transparent,
      seance: pallete.seance400_50,
      seanceFull: pallete.seance700,
    },
    dark: {
      primary: pallete.primaryBlack,
      secondary: pallete.primaryBlack,
      colored: pallete.transparent,
      transparent: pallete.transparent,
      seance: pallete.seance400_50,
      seanceFull: pallete.seance700,
    },
  },

  gradients: {
    main: ['#E79C1C', '#405c57ff', '#000000', '#000000'],
    mainDarker: ['#000000', '#000000', '#000000', '#405c57ff', '#E79C1C'],
    purpleTransparent: ['#000000f0', '#405c571a'],
    purpleToFullTransparent: ['transparent', '#405c57ff'],
    premium: [
      '#955D08',
      '#D19843',
      '#F5BC67',
      '#A16813',
      '#A16813',
      '#FEC672',
      '#D9A04B',
    ],
    redExit: [
      '#B84B4B',
      '#FFB9B9',
      '#D86B6B',
      '#D86B6B',
      '#AB3E3E',
      '#AB3E3E',
      '#D86B6B',
      '#BC5050',
    ],
    error: ['red', 'red'],
    blacToWhiteTransparent: ['#FFFFFF80', '#00000080'],
    blackToFullTransparent: ['transparent', '#00000080'],
  },
};
