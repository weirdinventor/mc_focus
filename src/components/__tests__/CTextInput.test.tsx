import { CTextInput } from '@components/CTextInput';
import { Colors } from '@constants/Colors';
import { store } from '@store/index';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';
import { Provider } from 'react-redux';

describe('[COMPONENTS]: CTextInput', () => {
  it('renders correctly', () => {
    const tree = render(<CTextInput placeholderText="test" />);

    const pressable = tree.getByRole('button');
    fireEvent.press(pressable);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders with label text', () => {
    const { getByText } = render(
      <Provider store={store}>
        <CTextInput labelText="test" />
      </Provider>,
    );
    const labelArea = getByText('test');

    expect(labelArea).toBeTruthy();
  });

  it('renders with right accessory', () => {
    const { getByText } = render(
      <Provider store={store}>
        <CTextInput RightAccessory={() => <Text>test</Text>} />
      </Provider>,
    );
    const rightAccessory = getByText('test');

    expect(rightAccessory).toBeTruthy();
  });

  it('renders with correct disabled color', () => {
    const { getByPlaceholderText, getByRole } = render(
      <Provider store={store}>
        <CTextInput placeholderText="test" editable={false} />
      </Provider>,
    );

    const pressable = getByRole('button');
    const input = getByPlaceholderText('test');

    fireEvent.press(pressable);

    expect(input).toHaveStyle({ color: Colors.primaryBlack });
  });
});
