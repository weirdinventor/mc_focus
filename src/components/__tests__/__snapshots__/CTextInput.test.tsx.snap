// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[COMPONENTS]: CTextInput renders correctly 1`] = `
<View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#e7e5e8",
            "borderColor": "#01080a",
            "borderRadius": 20,
            "borderWidth": 1,
            "flexDirection": "row",
            "overflow": "hidden",
            "paddingLeft": 24,
            "paddingRight": 16,
            "paddingVertical": 20,
          },
          false,
          undefined,
        ]
      }
    >
      <TextInput
        autoCapitalize="none"
        autoComplete="off"
        autoCorrect={false}
        editable={true}
        placeholder="test"
        placeholderTextColor="#e7e5e8"
        style={
          [
            {
              "alignSelf": "stretch",
              "flex": 1,
              "paddingHorizontal": 0,
              "paddingVertical": 0,
            },
            {
              "color": "#202020",
              "fontFamily": "GoogleSans-Bold",
              "fontSize": 20,
            },
            false,
            false,
            undefined,
          ]
        }
        textAlignVertical="center"
      />
    </View>
  </View>
</View>
`;
